diff --git a/apt/private/apt_deb_repository.bzl b/apt/private/apt_deb_repository.bzl
index c703cf1..6205e8b 100644
--- a/apt/private/version.bzl
+++ b/apt/private/version.bzl
@@ -132,7 +132,7 @@ def _compare_version(va, vb):
     return _version_cmp_part(vap[2] or "0", vbp[2] or "0")
 
 def _sort(versions, reverse = False):
-    vr = versions
+    vr = list(versions)
     for i in range(len(vr)):
         for j in range(i + 1, len(vr)):
             # if vr[i] is greater than vr[i+1] then swap their indices.
